File: 3. Random Number Generators/e3q3_fibonacci.m
%% Fibonacci random number generator (Knuth)
% Seydel, Course Notes, Chapter 2, pages 211-212

% Parameters
mu = 5;
nu = 17;
M = 714025;
%seed = 12345;
seed = mod(tic,M);
nsample = 10000;

% Linear congruential generator to start the sequence
a = 1366; b = 150889;
U = zeros(1,nu);
U(1) = seed;
for n = 2:nu
    U(n) = mod(a*U(n-1)+b,M);  
end

U

% % Fibonacci generator
% for n = nu+1:nsample
%     U(n) = mod(U(n-nu)-U(n-mu),M);  
% end
% 
% U = U/M;


% %% Tests
% 
% % Bin the random variables in a histogram and normalise it
% nbins = 100;
% [h,x] = hist(U,100);
% h = h*nbins/nsample; % normalisation
% close all
% 
% % Plot of the probability density function
% figure(1)
% plot(x,h,x,ones(size(x)))
% xlim([-0.2 1.2])
% ylim([0 2])
% xlabel('x')
% ylabel('f')
% legend('Sampled','Theory')
% title('Uniform distribution')
% 
% % 2D scatter plot
% figure(2)
% plot(U(1:nsample-1),U(2:nsample),'.')
% xlabel('U_{i-1}')
% ylabel('U_i')
% title('Scatter plot')
% 
% % 3D scatter plot
% figure(3)
% plot3(U(1:nsample-2),U(2:nsample-1),U(3:nsample),'.')
% xlabel('U_i')
% ylabel('U_{i+1}')
% zlabel('U_{i+2}')
% title('3D scatter plot')

File: 3. Random Number Generators/e3q3_fibonacci.py
import numpy as np
import matplotlib.pyplot as plt

def lcg_int(seed, a, b, M, n):
    """Linear Congruential Generator returning n INTEGER states in [0, M-1]."""
    x = seed
    out = []
    for _ in range(n):
        x = (a * x + b) % M
        out.append(x)
    return out

def fibonacci_rng(a, b, u, v, M, seed, nsamples, op="minus"):
    """
    Lagged-Fibonacci generator:
      X_n = (X_{n-u} ± X_{n-v}) mod M
    Returns nsamples uniforms U_n = X_n / M.
    """
    if not (1 <= u < v):
        raise ValueError("Require 1 <= u < v.")

    # 1) Get integer seeds X_0..X_{v-1} via LCG
    X = lcg_int(seed, a, b, M, v)

    # 2) Generate X_v .. X_{v+nsamples-1}
    for n in range(v, v + nsamples):
        if op == "minus":
            xn = (X[n - u] - X[n - v]) % M
        elif op == "plus":
            xn = (X[n - u] + X[n - v]) % M
        else:
            raise ValueError("op must be 'minus' or 'plus'.")
        X.append(xn)

    # 3) Normalise the last nsamples values only
    X_tail = np.array(X[v:v + nsamples], dtype=float)
    return X_tail / M  # U_n in [0,1)

def scatter_square(a, b, u, v, M, seed, nsamples=4000, op="minus"):
    U = fibonacci_rng(a, b, u, v, M, seed, nsamples, op=op)

    # Build (U_n, U_{n-1}) pairs
    x = U[1:]           # U_n
    y = U[:-1]          # U_{n-1}

    plt.figure(figsize=(8, 6))
    plt.plot(x, y, 'rx', markersize=2)
    plt.xlabel('U\u2099')                 # Uₙ
    plt.ylabel('U\u2099\u208B\u2081')     # Uₙ₋₁
    plt.title(f'Unit Square Scatter of Lagged Fibonacci '
              f'(nsamples={nsamples}, u={u}, v={v}, M={M}, op={op})')
    plt.tight_layout()
    plt.show()

# Example (tweak to match Seydel’s exact parameters)
scatter_square(a=1597, b=51749, u=5, v=17, M=644_980, seed=24_234, nsamples=4000, op="minus")